### PR TITLE
perf(ext/ffi): leverage V8 Fast Calls

### PR DESCRIPTION
**Pull request by littledivy**
100x faster calls into FFI

image

### CODE REVIEW

File: test_ffi/tests/bench.js [hunk 1]
```diff
@@ -220,228 +221,288 @@ const dylib = Deno.dlopen(libPath, {
...
+  return_u32();
 });
 
+const { return_i32 } = dylib.symbols;
 Deno.bench("return_i32()", () => {
-  dylib.symbols.return_i32();
+  return_i32();
 });
 
+const { return_u64 } = dylib.symbols;
```
Comment (bartlomieju) [inline test_ffi/tests/bench.js:330]: This is somewhat problematic, if you don't do this destructure we will never get optimized call, right?

**Review by bartlomieju — APPROVED** [review]
LGTM

File: ext/ffi/lib.rs [hunk 2]
```diff
@@ -623,13 +624,91 @@ where
...
+    }
+    fast_ffi_templ = Some(FfiFastCallTemplate {
+      args: args.into_boxed_slice(),
+      ret: (&fast_api::Type::from(&sym.result_type)).into(),
+      symbol_ptr: sym.ptr.as_ptr() as *const c_void,
+    });
+  }
+
+  let sym = Box::leak(sym);
+  let builder = v8::FunctionTemplate::builder(
```
Comment (aapoalas) [inline ext/ffi/lib.rs:711]: Changing from `v8::Function` to `v8::FunctionTemplate` means that the functions become un-collectable by the GC. Should there be a possible option to opt out from the optimization?

File: test_ffi/tests/bench.js (hunk 1 above)
Comment (littledivy) [reply test_ffi/tests/bench.js:330]: Yes, but it's not really problematic. The prototype chain probably prevents function optimization, a good benchmark / performant code will do this.

File: ext/ffi/lib.rs [hunk 3]
```diff
@@ -623,13 +624,91 @@ where
...
+      NativeType::U8 | NativeType::U16 | NativeType::U32 => {
+        fast_api::Type::Uint32
+      }
+      NativeType::I8 | NativeType::I16 | NativeType::I32 => {
+        fast_api::Type::Int32
+      }
+      NativeType::F32 => fast_api::Type::Float32,
+      NativeType::F64 => fast_api::Type::Float64,
+      NativeType::Void => fast_api::Type::Void,
+      NativeType::Function
```
Comment (aapoalas) [inline ext/ffi/lib.rs:660]: I _guess_ theoretically we could support `int64_t` and `uint64_t` argument types, meaning also pointers. The slow functions already have (or at least had) support for reading a number as a pointer value.  Allowing `uint64_t` to stand in here for pointers et al. would mean that users could opt in to the Fast Call API by explicitly using `Number(UnsafePointer.of(typedArray))` for pointer values. Of course, users would also need to take care to not lose any data through this.

File: ext/ffi/lib.rs [hunk 4]
```diff
@@ -623,13 +624,91 @@ where
...
+  if !sym.parameter_types.iter().any(|t| !is_fast_api(*t))
+    && is_fast_api(sym.result_type)
+  {
+    let mut args = sym
+      .parameter_types
+      .iter()
+      .map(|t| t.into())
+      .collect::<Vec<_>>();
+    if args.is_empty() {
+      args.push(fast_api::Type::V8Value);
```
Comment (aapoalas) [inline ext/ffi/lib.rs:701]: Does Fast API not support empty arguments?

Comment (littledivy) [reply ext/ffi/lib.rs:701]: Apparently needs the receiver V8Value for empty calls.

Comment (aapoalas) [reply ext/ffi/lib.rs:701]: Huh. That's odd and dumb. Did you happen to try with a single void argument? That would at least make a little sense with how C functions that absolutely do not accept parameters are declared with a single void argument.

Comment (littledivy) [reply ext/ffi/lib.rs:701]: Void arguments trigger unreachable paths in V8

**Pull request merged**